### ISSUE TITLE
New Rule Proposal: `no-input-type-text`

### ISSUE THREAD

**Issue by karlhorky** [yannickcr/eslint-plugin-react#2866]
Hi there 👋

Thanks for the continued work on this great plugin!

I had an idea for a rule today, and wanted to test the waters here on a `no-input-type-text` rule (because [the default value for `type` is `text` already](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#frame_exampletel:~:text=The%20default%20value,-.)):

Example of **incorrect** code:

```jsx
<input type="text" />
```

Example of **correct** code:

```jsx
<input />
```

**Comment by ljharb** [comment]
It's far, far better to always require explicitness, even when things have defaults. I'd rather see a rule that forbids `<input />` than one that requires people to keep an extra thing in their heads (that the default type is "text").

**Comment by karlhorky** [comment]
> It's far, far better to always require explicitness, even when things have defaults

I understand the mindset, and in some cases, I agree.

But under some circumstances, it's nice to not have to write the verbose thing. And if you have a lot of code with `input`s, this will reduce the amount of code that needs to be written / reviewed.

Also, and maybe more importantly, this is an opinion - other individuals and teams will have different ones :)

So I guess my question would be: why not both? There's a lot of prior art of ESLint rules that offer options for both opinions.

**Comment by ljharb** [comment]
I don't think that just "input" is sufficient to warrant an entire rule. There's also already a rule, jsx-a11y/button-has-type, for button elements, but that's because the default is unintuitive.

Are there other HTML elements that this would apply to, where the default can be omitted and also is unsurprising? A single rule that encompasses a bunch of HTML elements in this way, autofixable, with both options, seems like it might be worth the complexity.

**Comment by karlhorky** [comment]
> Are there other HTML elements that this would apply to, where the default can be omitted and also is unsurprising?

Looks like `<ol>` and the `<li>` elements contained within have `type` attributes:

> - a for lowercase letters
> - A for uppercase letters
> - i for lowercase Roman numerals
> - I for uppercase Roman numerals
> - 1 for numbers (default)

[`<ol>` on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ol#Attributes)

The `<ul>` element also has a `type` attribute, but this attribute is deprecated.

Also (but maybe less interestingly):

- `<style>` - default: `type="text/css"`
- `<script>` - default: `type="text/javascript"`

**Comment by ljharb** [comment]
huh, i've never heard of those being used - only the CSS.

In React, `<script>` doesn't really make sense to address. So that would basically leave us with `button`, `input`, and `style` - not all that many :-/

**Comment by karlhorky** [comment]
Hmm... and what's your opinion on other default props / HTML attributes? Things that can be omitted and the functionality is the same?

**Comment by ljharb** [comment]
I think the longer a list we build up, the more convincing a rule for it becomes.

**Comment by karlhorky** [comment]
Ok, let me do a bit of research, and I will come back with a longer list. I'll change the issue title and description in the meantime.

**Comment by karlhorky** [comment]
Ok, so I've updated the list above with more examples.

Talking with @wooorm, it seems like there are some interesting resources in the `rehype-minify-enumerated-attribute` package which could help with finding the attributes with default values.

I've asked if publishing a separate package with this data would be possible at https://github.com/rehypejs/rehype-minify/issues/37

**Comment by ljharb** [comment]
I have https://npmjs.com/html-element-map, which won’t help here but is a similar concept re extracted data as a package, so that would likely work well.

**Comment by wooorm** [comment]
Yeah I’m fine publishing it separately! The Q then is: what format do you need? Does the current schema make sense?

Another thing is that the schema currently uses hast casing, which is in rare cases different from react casing (although the translation is [relatively simple](https://github.com/wooorm/property-information/blob/main/hast-to-react.json)), and both are different from HTML.

**Comment by ljharb** [comment]
Note that publishing it separately is most compelling with other projects can start using it - that way it ensures it will stay correct :-)

I have no idea what "hast" is, but it seems reasonable for the package to provide html casing, and for consumers to adapt it as needed (or, for another package to wrap it and provide the react version, eg)

**Comment by wooorm** [comment]
> I have no idea what "hast" is, but it seems reasonable for the package to provide html casing, and for consumers to adapt it as needed (or, for another package to wrap it and provide the react version, eg)

[hast](https://github.com/syntax-tree/hast) is an html ast. Like estree but for HTML. Sometimes but not always used with [mdast](https://github.com/syntax-tree/mdast) (markdown).
It’s relevant here because the HTML minifier in question uses hast.

Transformation between react / hast / html / dom is all not very complex, but does need a bit of a look up table. As we’re currently interested in react / hast (almost identical, close to DOM props), using that sounds like the simplest path forward to me?

Anyway, I’m not blocking this, just preferences

**Comment by ljharb** [comment]
Ideally I'd like this plugin to use something that's very simple and barebones - just DOM to react jsx - that shares a common core with what the minifier uses, rather than using something multi-purpose that will have more version churn. Is that a possibility?

**Comment by karlhorky** [comment]
Some more discussion about publishing the new package over here: https://github.com/rehypejs/rehype-minify/issues/37

**Comment by karlhorky** [comment]
Created a rudimentary version to prevent `<input type="text" />` and `<button type="submit"><button>`:

Code: https://github.com/upleveled/eslint-plugin-upleveled/blob/main/rules/no-unnecessary-html-attributes.ts
Docs: https://github.com/upleveled/eslint-plugin-upleveled/blob/main/docs/rules/no-unnecessary-html-attributes.md

First plugin rule I've written, wild ride working with the AST! 🙌 And also finding the types for TypeScript.

**Comment by ljharb** [comment]
@karlhorky i'm still hopeful this can rule can be upstreamed here, once there's a core package shared by us and hast.
